### PR TITLE
planner: prepared plan cache support limit

### PR DESCRIPTION
**Pull request by fzzf678**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### What is changed, added or deleted? (Required)
Starting from v6.6, plan cache support supports limit.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v6.6 (TiDB 6.6 versions)
- [ ] v6.5 (TiDB 6.5 versions)
- [ ] v6.4 (TiDB 6.4 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):
https://github.com/pingcap/tidb/issues/40219
### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

**Review by Oreoxmt — APPROVED** [review]
LGTM

**Comment by TomShawn** [comment]
According to @qw4990, this feature might be postponed. cc @shichun-0415

**Comment by shichun-0415** [comment]
> According to @qw4990, this feature might be postponed. cc @shichun-0415

According to information from @songrijie, it is correct that plan cache supports limit in v6.6. The support for subqueries will be released in future sprints.

File: system-variables.md [hunk 1]
```diff
@@ -1503,6 +1503,12 @@ MPP 是 TiFlash 引擎提供的分布式计算框架，允许节点之间的数
 - 默认值：`ON`
 - 该变量指定是否对窗口函数采用流水线的执行算法。
 
+### `tidb_enable_plan_cache_for_param_limit`
```
Comment (Oreoxmt) [inline system-variables.md:1506]:
```suggestion
### `tidb_enable_plan_cache_for_param_limit` <span class="version-mark">从 v6.6.0 版本开始引入</span>

```

File: system-variables.md [hunk 2]
```diff
@@ -1503,6 +1503,12 @@ MPP 是 TiFlash 引擎提供的分布式计算框架，允许节点之间的数
...
+- 作用域：SESSION | GLOBAL
+- 是否持久化到集群：是
```
Comment (Oreoxmt) [inline system-variables.md:1508]:
```suggestion
- 是否持久化到集群：是
- 类型：布尔型
```

File: system-variables.md [hunk 3]
```diff
@@ -1503,6 +1503,12 @@ MPP 是 TiFlash 引擎提供的分布式计算框架，允许节点之间的数
...
+- 默认值：`ON`
+- 这个变量用来控制 Prepared Plan Cache 是否缓存 `limit` 后带有参数的执行计划。
+
```
Comment (fzzf678) [inline system-variables.md:1511]: @TomShawn supply description of a new switch, PTAL

File: system-variables.md [hunk 4]
```diff
@@ -1503,6 +1503,14 @@ MPP 是 TiFlash 引擎提供的分布式计算框架，允许节点之间的数
 - 默认值：`ON`
 - 该变量指定是否对窗口函数采用流水线的执行算法。
 
+### `tidb_enable_plan_cache_for_param_limit` <span class="version-mark">从 v6.6.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 是否持久化到集群：是
+- 类型：布尔型
+- 默认值：`ON`
+- 这个变量用来控制 Prepared Plan Cache 是否缓存 `limit` 后带有参数的执行计划。
```
Comment (shichun-0415) [inline system-variables.md:1512]: 参数值不大于 10000 这个条件要加上吗？

Comment (fzzf678) [inline system-variables.md:1512]:
```suggestion
- 这个变量用来控制 Prepared Plan Cache 是否缓存 `limit` 后带有参数的执行计划， `Limit` 后面的 `Count` 具体参数值大于 10000 的执行计划目前不支持缓存。
```

Comment (fzzf678) [reply system-variables.md]: fixed, PTAL again

Comment (shichun-0415) [reply system-variables.md]: 根据上面文件的 '大于 10000 的执行计划不缓存' 这个结论，这个变量控制的是不是小于 10000 的执行计划的缓存？如果是这样，这里可能要改成：
> 这个变量用来控制 Prepared Plan Cache 是否缓存 `limit` 后带有 `Count` 参数且参数值小于 10000 的执行计划。

**Review by shichun-0415 — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]
LGTM

File: sql-prepared-plan-cache.md [hunk 5]
```diff
@@ -22,7 +22,7 @@ TiDB 优化器对这两类查询的处理是一样的：`Prepare` 时将参数
 - 包含 `ignore_plan_cache` 这一 Hint 的查询，例如 `select /*+ ignore_plan_cache() */ * from t`；
 - 包含除 `?` 外其他变量（即系统变量或用户自定义变量）的查询，例如 `select * from t where a>? and b>@x`；
 - 查询包含无法被缓存函数。目前不能被缓存的函数有：`database()`、`current_user`、`current_role`、`user`、`connection_id`、`last_insert_id`、`row_count`、`version`、`like`；
-- `?` 在 `Limit` 后的查询，如 `Limit ?` 或者 `Limit 10, ?`，此时 `?` 的具体值对查询性能影响较大，故不缓存；
+- `Limit` 后面的 `Count` 为 `?` 且具体参数值大于 10000 的执行计划不缓存；
```
Comment (shichun-0415) [inline sql-prepared-plan-cache.md:25]:
```suggestion
- `Limit` 后面带一个变量 (`LIMIT ?`) 且变量值大于 10000 的执行计划不缓存；
```

File: system-variables.md [hunk 6]
```diff
@@ -1503,6 +1503,14 @@ MPP 是 TiFlash 引擎提供的分布式计算框架，允许节点之间的数
 - 默认值：`ON`
 - 该变量指定是否对窗口函数采用流水线的执行算法。
 
+### `tidb_enable_plan_cache_for_param_limit` <span class="version-mark">从 v6.6.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 是否持久化到集群：是
+- 类型：布尔型
+- 默认值：`ON`
+- 这个变量用来控制 Prepared Plan Cache 是否缓存 `Limit` 后带有参数的执行计划。目前不支持缓存 `Limit` 后面的 `Count` 具体参数值大于 10000 的执行计划。
```
Comment (shichun-0415) [inline system-variables.md:1512]:
```suggestion
- 这个变量用来控制 Prepared Plan Cache 是否缓存 `Limit` 后面带变量 (`LIMIT ?`) 的执行计划。目前不支持缓存 `Limit` 后面带变量且变量值大于 10000 的执行计划。
```

File: sql-prepared-plan-cache.md (hunk 5 above)
Comment (shichun-0415) [reply sql-prepared-plan-cache.md:25]: @qw4990 @Oreoxmt @fzzf678 PTAL

File: system-variables.md (hunk 6 above)
Comment (shichun-0415) [reply system-variables.md:1512]: @qw4990 @Oreoxmt @fzzf678 PTAL

File: sql-prepared-plan-cache.md (hunk 5 above)
Comment (Oreoxmt) [reply sql-prepared-plan-cache.md:25]: LGTM

File: system-variables.md (hunk 6 above)
Comment (Oreoxmt) [reply system-variables.md]:
```suggestion
- 这个变量用来控制 Prepared Plan Cache 是否缓存 `LIMIT` 后面带变量 (`LIMIT ?`) 的执行计划。目前不支持缓存 `LIMIT` 后面带变量且变量值大于 10000 的执行计划。
```
Rest LGTM

File: sql-prepared-plan-cache.md [hunk 7]
```diff
@@ -22,7 +22,7 @@ TiDB 优化器对这两类查询的处理是一样的：`Prepare` 时将参数
 - 包含 `ignore_plan_cache` 这一 Hint 的查询，例如 `select /*+ ignore_plan_cache() */ * from t`；
 - 包含除 `?` 外其他变量（即系统变量或用户自定义变量）的查询，例如 `select * from t where a>? and b>@x`；
 - 查询包含无法被缓存函数。目前不能被缓存的函数有：`database()`、`current_user`、`current_role`、`user`、`connection_id`、`last_insert_id`、`row_count`、`version`、`like`；
-- `?` 在 `Limit` 后的查询，如 `Limit ?` 或者 `Limit 10, ?`，此时 `?` 的具体值对查询性能影响较大，故不缓存；
+- `Limit` 后面带一个变量 (`LIMIT ?`) 且变量值大于 10000 的执行计划不缓存；
```
Comment (Oreoxmt) [inline sql-prepared-plan-cache.md:25]:
```suggestion
- `LIMIT` 后面带一个变量 (`LIMIT ?`) 且变量值大于 10000 的执行计划不缓存；
```

File: system-variables.md [hunk 8]
```diff
@@ -1503,6 +1503,14 @@ MPP 是 TiFlash 引擎提供的分布式计算框架，允许节点之间的数
 - 默认值：`ON`
 - 该变量指定是否对窗口函数采用流水线的执行算法。
 
+### `tidb_enable_plan_cache_for_param_limit` <span class="version-mark">从 v6.6.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 是否持久化到集群：是
+- 类型：布尔型
+- 默认值：`ON`
+- 这个变量用来控制 Prepared Plan Cache 是否缓存 `Limit` 后面带变量 (`LIMIT ?`) 的执行计划。目前不支持缓存 `Limit` 后面带变量且变量值大于 10000 的执行计划。
```
Comment (Oreoxmt) [inline system-variables.md:1512]:
```suggestion
- 这个变量用来控制 Prepared Plan Cache 是否缓存 `LIMIT` 后面带变量 (`LIMIT ?`) 的执行计划。目前不支持缓存 `LIMIT` 后面带变量且变量值大于 10000 的执行计划。
```

**Review by Oreoxmt — APPROVED** [review]

**Pull request merged**